### PR TITLE
Add mc admin tier subcommands

### PR DESCRIPTION
**Pull request by krisis**

### CODE REVIEW

File: cmd/admin-storage-class-add.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+		Usage:  "gcs credentials file used to access/transition objects to GCS",
+		Hidden: true,
+	},
+}
+
+var adminStorageClassAddFlags = []cli.Flag{
+	cli.StringFlag{
+		Name:  "type",
+		Value: "",
+		Usage: "cloud backend type. Valid options are '[s3, azure, gcs]'",
```
Comment (poornas) [inline cmd/admin-storage-class-add.go:53]: should also include minio  :)

File: cmd/admin-storage-class-add.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+  Supported values are s3, azure or gcs.
+
+CLOUD_BACKEND_FLAGS:
+  Cloud storage backend specific flags.
+
+FLAGS:
+  {{range .VisibleFlags}}{{.}}
+  {{end}}
+EXAMPLES:
+  1. Configure a new storage class backed by AWS s3 using its Standard-IA storage class
```
Comment (poornas) [inline cmd/admin-storage-class-add.go:148]:
```suggestion
  1. Configure a new storage class backend by AWS s3 using its Standard-IA storage class
```

File: cmd/admin-storage-class-add.go [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+		}
+
+		s3SC := ctx.String("storage-class")
+		if s3SC != "" {
+			s3Opts = append(s3Opts, madmin.S3StorageClass(s3SC))
+		}
+
+		s3Cfg, err := madmin.NewTransitionStorageClassS3(scName, accessKey, secretKey, bucket, s3Opts...)
+		if err != nil {
+			fatalIf(probe.NewError(err), "Invalid configuration for s3 backed storage-class")
```
Comment (poornas) [inline cmd/admin-storage-class-add.go:202]:
```suggestion
			fatalIf(probe.NewError(err), "Invalid configuration for s3 transition target")
```

File: cmd/admin-storage-class-add.go [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+		}
+
+		prefix := ctx.String("prefix")
+		if prefix != "" {
+			azOpts = append(azOpts, madmin.AzurePrefix(prefix))
+		}
+
+		azCfg, err := madmin.NewTransitionStorageClassAzure(scName, accessKey, secretKey, bucket, azOpts...)
+		if err != nil {
+			fatalIf(probe.NewError(err), "Invalid configuration for azure backed storage-class")
```
Comment (poornas) [inline cmd/admin-storage-class-add.go:234]:
```suggestion
			fatalIf(probe.NewError(err), "Invalid configuration for azure backend storage-class")
```

File: cmd/admin-storage-class-add.go (hunk 3 above)
Comment (poornas) [reply cmd/admin-storage-class-add.go:202]: `s3 backend storage-class` will be misleading and get confused with actual storage class used on s3

File: cmd/admin-storage-class-add.go [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+		if err != nil {
+			fatalIf(probe.NewError(err), "Invalid configuration for gcs backed storage-class")
+		}
+
+		return madmin.StorageClassConfig{
+			Type: madmin.Gcs,
+			Gcs:  gcsCfg,
+		}
+	}
+	return madmin.StorageClassConfig{}
```
Comment (poornas) [inline cmd/admin-storage-class-add.go:274]: the command line should also take the target storage class where object should be tiered to. Seems to be missing for all 3 clouds

**Review by poornas — CHANGES_REQUESTED** [review]

**Comment by poornas** [comment]
can you fix conflicts here @krisis

File: cmd/admin-tier-add.go [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+	},
+	cli.StringFlag{
+		Name:  "prefix",
+		Value: "",
+		Usage: "cloud backend target prefix",
+	},
+	cli.StringFlag{
+		Name:  "storage-class",
+		Value: "",
+		Usage: "AWS S3 or compatible object storage's storage-class. e.g AWS s3 suppports S3_STANDARD and S3_IA",
```
Comment (ravindk89) [inline cmd/admin-tier-add.go:80]: As discussed, should modify these to specify only supported storage classes:

- `STANDARD` , `REDUCED_REDUNDANCY` <- via [S3 spec](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#AmazonS3-PutObject-request-header-StorageClass).

File: cmd/admin-tier-add.go [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+  Tier type specific flags.
+
+FLAGS:
+  {{range .VisibleFlags}}{{.}}
+  {{end}}
+EXAMPLES:
+  1. Configure a new remote tier which transitions objects to Azure Blob Storage.
+     {{.Prompt}} {{.HelpName}} azure myminio AZTIER --account-name foobar --account-key foobar123 --region us-east-1 --bucket testbucket --prefix testprefix/
+
+  2. Configure a new remote tier which transitions objects to AWS S3 Standard-IA storage class.
```
Comment (ravindk89) [inline cmd/admin-tier-add.go:113]: @krisis Need to change the `--storage-class` value here. Also probably referencing `REDUCED_REDUNDANCY` unless we're sure we can support `STANDARD_IA`

**Review by ravindk89 — APPROVED** [review]
So from a general perspective this looks good to me. The messaging I think is fine and the args look good.

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**